infra/aws/terraform/registry.k8s.io/s3/bucket.tf

```
@@ -70,3 +70,48 @@ resource "aws_s3_bucket_ownership_controls" "registry-k8s-io" {
     aws_s3_bucket_policy.registry-k8s-io-public-read
   ]
 }
+
+# Versioning must be enabled for S3 replication
+resource "aws_s3_bucket_versioning" "registry-k8s-io" {
+  bucket = aws_s3_bucket.registry-k8s-io.id
+  versioning_configuration {
+    status = "Enabled"
+  }
+}
+
+resource "aws_s3_bucket_replication_configuration" "registry-k8s-io" {
+  count = var.s3_replication_iam_role_arn != "" && length(var.s3_replication_rules) > 0 ? 1 : 0
+
+  # Must have bucket versioning enabled first
+  depends_on = [aws_s3_bucket_versioning.registry-k8s-io]
+
+  role = var.s3_replication_iam_role_arn
+
+  bucket = aws_s3_bucket.registry-k8s-io.id
+
+  dynamic "rule" {
+    for_each = var.s3_replication_rules
+
+    content {
+      id = rule.value.id
+
+      status = rule.value.status
+
+      # Set priority, filter and delete_marker_replication to use V2 schema for multiple 
+      # destination bucket rules
+      priority = rule.value.priority
+
+      filter {}
+
+      delete_marker_replication {
+        status = "Enabled"
+      }
+
+      destination {
+        bucket        = rule.value.destination_bucket_arn
+        storage_class = rule.value.destination_bucket_storage_class
```

This might be needed

```suggestion
        storage_class = rule.value.destination_bucket_storage_class
                metrics {
          event_threshold {
            minutes = 15
          }
          status = "Enabled"
        }
        replication_time {
          status = "Enabled"
          time {
            minutes = 15
          }
        }
      }
```

---

I just want to mention that enabling [RTC](https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication-time-control.html) changes the data transfer price rate and metrics are billed as custom metrics in cloudwatch. If we are good with the  budget it'll be a great idea to enable it. 

S3 Replication Time Control data transfer†	$0.015 per GB

---

If not time based, when would the syncing take place?

---

I tested only us regions. 
us-west-2 and us-east-1 replicated objects ~ 5 minutes or less. 
us-west-1 replication took 5 - 10 minutes.

This is what AWS support says about S3 replication time.

https://aws.amazon.com/premiumsupport/knowledge-center/s3-object-replication-time/

---

Within the same day should be fine

---

We can enable RTC any time if necessary.